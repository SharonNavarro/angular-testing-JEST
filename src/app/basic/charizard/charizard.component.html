<h1>Pok√©mon</h1>
@if (pokemon$ | async; as charizard) {
  <h3 data-test="h3-pokemon-name">{{charizard?.name | titlecase }}</h3>
  <img data-test="h3-pokemon-img" [src]="charizard?.sprites?.front_default" [alt]="charizard?.name">
} @else {
  <h2 data-test="h2-loading">Loading...</h2>
}
