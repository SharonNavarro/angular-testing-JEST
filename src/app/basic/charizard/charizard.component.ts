import { Component, inject, OnInit } from '@angular/core';
import { Observable } from 'rxjs';

import { Pokemon } from '../../shared/interfaces/pokemon.interface';
import { PokemonService } from '../../shared/services/pokemon.service';

@Component({
  selector: 'app-charizard',
  templateUrl: './charizard.component.html',
  styleUrl: './charizard.component.scss',
})
export class CharizardComponent implements OnInit {
  private _pokemonService = inject(PokemonService);
  public pokemon$!: Observable<Pokemon>;

  ngOnInit(): void {
    this.getPokemon();
  }

  private getPokemon(): void {
    this.pokemon$ = this._pokemonService.getPokemon(6);
  }
}
