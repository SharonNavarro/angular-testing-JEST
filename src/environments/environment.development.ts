export const environment = {
  POKEMON_URL: 'https://pokeapi.co/api/v2/pokemon',
};
