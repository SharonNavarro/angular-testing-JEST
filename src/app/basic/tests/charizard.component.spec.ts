import { AsyncPipe } from '@angular/common';
import { provideHttpClient } from '@angular/common/http';
import {
  HttpTestingController,
  provideHttpClientTesting,
} from '@angular/common/http/testing';
import {
  ComponentFixture,
  fakeAsync,
  TestBed,
  tick,
} from '@angular/core/testing';

import { environment } from '../../../environments/environment';
import { PokemonService } from '../../shared/services/pokemon.service';
import { CharizardComponent } from '../charizard/charizard.component';

describe('CharizardComponent', () => {
  let component: CharizardComponent;
  let fixture: ComponentFixture<CharizardComponent>;
  let compiled: HTMLElement;
  let httpMock: HttpTestingController;
  const urlService = environment.POKEMON_URL;

  beforeEach(async () => {
    await TestBed.configureTestingModule({
      declarations: [CharizardComponent],
      imports: [AsyncPipe],
      providers: [
        provideHttpClient(),
        provideHttpClientTesting(),
        PokemonService,
      ],
    }).compileComponents();
  });

  beforeEach(() => {
    fixture = TestBed.createComponent(CharizardComponent);
    component = fixture.componentInstance;
    httpMock = TestBed.inject(HttpTestingController);

    fixture.detectChanges();
    compiled = fixture.nativeElement;
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should match to snapshot', () => {
    expect(compiled.innerHTML).toMatchSnapshot();
  });

  it('should display a loading at the beginning', () => {
    const h2 = compiled.querySelector('[data-test="h2-loading"]');
    expect(h2?.textContent).toEqual('Loading...');
  });

  it('should display Charizard data correctly', fakeAsync(() => {
    const dummyPokemon = {
      name: 'charixado!!',
      sprites: {
        front_default: 'https://charizard.com/sprite.png',
      },
    };

    const request = httpMock.expectOne(`${urlService}/6`);
    expect(request.request.method).toBe('GET');

    request.flush(dummyPokemon);

    tick();
    fixture.detectChanges();

    const h3 = compiled.querySelector('[data-test="h3-pokemon-name"]');
    const img = compiled.querySelector(
      '[data-test="h3-pokemon-img"]'
    ) as HTMLImageElement;

    expect(h3?.textContent?.toLowerCase()).toContain(
      dummyPokemon.name.toLowerCase()
    );
    expect(img?.src).toBe(dummyPokemon.sprites.front_default);
    expect(img?.alt).toBe(dummyPokemon.name);
  }));
});
