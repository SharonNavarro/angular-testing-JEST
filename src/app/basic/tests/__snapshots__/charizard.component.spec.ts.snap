// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CharizardComponent should match to snapshot 1`] = `"<h1>Pok√©mon</h1><!--container--><h2 data-test="h2-loading">Loading...</h2><!--container-->"`;
